[PM-8457] [PM-8608] Members page - remove paging logic / fix search

eliykat opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->
https://bitwarden.atlassian.net/browse/PM-8457
## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

Remove old paging logic and replace with virtual scroll. This should also fix problems with the members page not loading / search not working / toggles not working.

**Changes**
- duplicate `base.people.component.ts` to `new-base.people.component.ts` - as the current base component is also used by the provider portal, and I don't want to update that yet. This will eventually replace the current base component once the provider portal is updated.
- remove all paging logic and use the component library table + virtual scroll instead. We already used the CL table elements, but we weren't using virtual scroll or its built-in search/filter functionality.
- other minor tweaks: changes to names to make things clearer, adding comments

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->

https://github.com/bitwarden/clients/assets/31796059/956522fd-4c7e-43cf-a3a1-4908acbfe5a5

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

eliykat commented on apps/web/src/app/admin-console/organizations/members/people.component.html:
```diff
@@ -63,258 +56,262 @@
+                "all" | i18n
+              }}</label>
+            </th>
+            <th bitCell bitSortable="email" default>{{ "name" | i18n }}</th>
```
I'm waiting for product/design advice on this - the column is called "name" but is sorted by email, which isn't clear to the user. I had previously sorted by `name || email` but I reverted those changes when I realised that wasn't the current behaviour.

eliykat replied:
This is OK for now!

vincentsalucci commented on apps/web/src/app/admin-console/organizations/members/people.component.html:
```diff
@@ -45,16 +45,9 @@
-  <p *ngIf="!searchedUsers.length">{{ "noMembersInList" | i18n }}</p>
-  <ng-container *ngIf="searchedUsers.length">
+<ng-container *ngIf="!loading">
+  <p *ngIf="!dataSource.data.length">{{ "noMembersInList" | i18n }}</p>
```
This isn't working as intended. I believe we should be accessing the `dataSource.filteredData.length` as the catalyst for whether to show/hide.

This would involve changing the `ng-container: *ngIf*` conditional below as well.

eliykat replied:
Great pick up!

eliykat commented:
Evidence of fixed functionality:

![Screenshot 2024-06-10 at 7 15 12 AM](https://github.com/bitwarden/clients/assets/31796059/384f91ce-90ec-40a9-9bf7-f13d92193448)

Hinton commented:
@eliykat I was looking into this for https://github.com/bitwarden/clients/pull/7981 but I ran into some issues with the rendering of the last element. I can't remember if that was for vault items only or also for admin / provider views.

eliykat commented:
There is some problem as I get to the end of a table that exceeds a single viewport. Double scrollbars appear, which can have the effect of obscuring the last item unless you scroll down in the inner scrollable area. Not sure if that's what you came across but it's a problem either way. Looking into it further.

https://github.com/bitwarden/clients/assets/31796059/1537f40e-b60f-45c4-8668-c5c0beea061e

eliykat commented:
Fixed, using a workaround that we use elsewhere in the app (see comment on https://github.com/bitwarden/clients/pull/9515/commits/d7380f18d09ca531fe9b6f24ea9294796553bd9c). I've raised it for discussion on Slack but I think it's good for now.

https://github.com/bitwarden/clients/assets/31796059/b99bb551-2dee-4cd1-8d59-911eaf64cc36

eliykat commented:
I have fixed 2 bugs from QA:

1. when closing a bulk action, the whole table briefly flashed white ([Jira ticket](https://bitwarden.atlassian.net/browse/PM-8804)). This is fixed by **not** setting `loading = true` when `load()` is called after the initial page load. It's better to continue to show the old data until the new data is available. This is also how the current base component is implemented. Fixed by https://github.com/bitwarden/clients/pull/9515/commits/4a7699dfcdde62455f7ec04aa71f3952fb7f415e, which also renames this to `firstLoaded` to better reflect the actual usage.

2. when opening a dialog, the content of the page jumps around. This appears to occur wherever virtual scroll + the new modalService are used together, such as with the collection dialog in the individual vault. (Note the "edit item" dialog doesn't use the new service):

https://github.com/bitwarden/clients/assets/31796059/507d6470-ca1a-4028-835c-c430669a4ed8

This is caused by our styling on `.cdk-global-scrollblock` in `tw-theme.css`. That class is applied to the html element when a dialog is opened, as part of a strategy that prevents the user from scrolling while the dialog is open. Changing `relative` to `static` appears to fix the issue, see also [this issue](https://github.com/angular/components/issues/7390) for discussion and in particular [this workaround](https://github.com/angular/components/issues/7390#issuecomment-486947151). However I don't fully understand what's going on here, so @willmartian can I please have your expertise. I see you've [tweaked it before](https://github.com/bitwarden/clients/pull/8151) so maybe you have some insight here.

One thing to note is that this is a hotfix (if we can help it) so it won't go through a full regression suite before release - I am pretty nervous about changing global css in these circumstances.

Video after these changes:

https://github.com/bitwarden/clients/assets/31796059/75cf89dc-31d0-4cec-9d4c-041c801ec42a

willmartian commented:
> when closing a bulk action, the whole table briefly flashed white ([Jira ticket](https://bitwarden.atlassian.net/browse/PM-8804)). This is fixed by not setting loading = true when load() is called after the initial page load. It's better to continue to show the old data until the new data is available. This is also how the current base component is implemented. Fixed by https://github.com/bitwarden/clients/commit/4a7699dfcdde62455f7ec04aa71f3952fb7f415e, which also renames this to firstLoaded to better reflect the actual usage.

I agree with `It's better to continue to show the old data until the new data is available`. I think ideally we should also show a semi-transparent loading scrim _over_ the table data, but that is probably out of scope because it isn't supported natively by `bit-table` (CL ticket created [here](https://bitwarden.atlassian.net/browse/CL-311)). 

❓ Why is the data being loading when the member dialog is cancelled? My expectation is that it would only load on update.

> when opening a dialog, the content of the page jumps around ([Jira ticket](https://bitwarden.atlassian.net/browse/PM-8811)). This appears to occur wherever virtual scroll + the new modalService are used together, such as with the collection dialog in the individual vault.

Yeah, +1 to your diagnosis. I [previously changed](https://github.com/bitwarden/clients/pull/8151) the scroll block to apply `position: relative` to fix a bug with select overlays being mispositioned within dialogs, and your change causes a regression there. Sadly, some of our dependencies make conflicting assumptions about styles on `body` and `html`, and as such I also worry about introducing bugs elsewhere. With that in mind, I would suggest passing in `NoopScrollingStrategy` when opening this dialog, and the CL team will pursue a fix to this more generally that will go through a more rigorous QA check.
